# PGA files study

On this notebook, we'll do some descriptive statistics not on PGA _index_ file, but on the siva files themselves. We want to understand avg/mean/min/max repository size, number of blobs per-repository, avg contribution per-user, etc.

The steps I intend to complete to reach the results are:

1. Download siva files (following [PGA documentation](https://github.com/src-d/datasets/tree/master/PublicGitArchive/pga))
2. Extract siva files (following [siva documentation](https://github.com/src-d/go-siva))
3. Query repos using gitbase (following [source{d} documentation](https://docs.sourced.tech/intro/#analyzing-git-repositories))


## Step 1 - Download siva files

We will use the terminal to download the pertinent siva files. I will follow the same criterion as the study on PGA index, so we are interested on repos that has **Jupyter Notebook** files only.

```
$ pga list --lang "Jupyter Notebook" -f csv > repos_jupyter.csv
```

This will give us as output a csv file with the repo's URL, siva filenames for the repo, languages, and much more information. Let's see how the csv looks like.

_**NOTE**: since the export of pga doesn't come with the headers on the first row, I manually added on pandas dataframe using the ones from the index, as they are the same._

In [31]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("repos_jupyter.csv", names=["URL", "SIVA_FILENAMES", "FILE_COUNT", "LANGS", "LANGS_BYTE_COUNT",
                                             "LANGS_LINES_COUNT", "LANGS_FILES_COUNT", "COMMITS_COUNT",
                                             "BRANCHES_COUNT", "FORK_COUNT", "EMPTY_LINES_COUNT", "CODE_LINES_COUNT",
                                             "COMMENT_LINES_COUNT", "LICENSE"])
df.describe()

FILE_COUNT  COMMITS_COUNT  BRANCHES_COUNT   FORK_COUNT
count   2606.000000    2606.000000     2606.000000  2606.000000
mean     471.105909     851.153108      142.298542     1.209133
std     2274.210560    3229.955311      677.735284     7.520236
min        1.000000       1.000000        2.000000     0.000000
25%       25.000000      27.000000        3.000000     0.000000
50%       73.000000      95.000000        9.000000     0.000000
75%      257.750000     415.000000       48.000000     0.000000
max    70056.000000   79104.000000    14709.000000    67.000000

In [32]:
df.head()

URL  \
0         https://github.com/statsmodels/statsmodels   
1  https://github.com/ptwobrussell/Mining-the-Soc...   
2               https://github.com/sagemath/sagecell   
3                    https://github.com/zeromq/pyzmq   
4               https://github.com/PyTables/PyTables   

                                      SIVA_FILENAMES  FILE_COUNT  \
0      0b9745d5b485ed8bd4a548543257c960da396d1a.siva        1657   
1      cafea358122f631f15925e253917717f45539e18.siva         144   
2  7ac27a0dc245243acfee08840782d56e954928a7.siva,...         237   
3  08c4dbb82205f6dbb3e3e1c024a10741bdc632d3.siva,...         290   
4      fe6189a37e94ce706e79985ca5d6ff91068ba960.siva         653   

                                               LANGS  \
0  AGS Script,Batchfile,C,CSS,CSV,Cython,HTML,JSO...   
1  CSS,HTML,JavaScript,Jupyter Notebook,Markdown,...   
2  Adobe Font Metrics,Batchfile,CSS,HTML,INI,JSON...   
3  Batchfile,C,C++,CSS,Cython,Dockerfile,HTML,Jup...   
4  Batchfile,C,C++,CMake,CSV,Cython,Diff,Gnuplot,...   

                                    LANGS_BYTE_COUNT  \
0  457842,7825,12088,30189,12522370,179965,155105...   
1     1179,21995,1619294,85048,17008,300682,1520,623   
2  557,4537,63637,182720,542,228,486271,5198,6871...   
3  3502,22766,33292,8920,112006,267,663,167790,72...   
4  13010,2860411,121115,35579,22668,285193,1999,2...   

                                   LANGS_LINES_COUNT  \
0  6985,249,432,478,194340,4209,2670,242,536,1655...   
1                   75,690,37629,2031,162,9528,28,24   
2  21,171,2550,4610,31,15,13721,224,187,1232,151,...   
3  91,1066,1256,553,3560,10,24,507,229,444,21208,...   
4  417,77456,3379,941,1603,8694,66,57,16,31,847,5...   

                                   LANGS_FILES_COUNT  COMMITS_COUNT  \
0  1,5,3,5,164,19,13,2,3,41,1,3,2,919,29,1,4,1,17...           9727   
1                                 2,17,12,3,4,96,1,2            168   
2          1,1,22,18,2,1,32,1,2,8,5,49,13,3,3,6,1,13           2436   
3            1,13,5,1,22,1,1,1,3,12,172,1,1,2,6,2,29           2361   
4  4,148,12,9,5,12,3,2,1,2,1,7,2,1,2,173,34,12,27...           3581   

   BRANCHES_COUNT  FORK_COUNT  \
0            2368           0   
1              22           0   
2             439           0   
3             799           0   
4             321           0   

                                   EMPTY_LINES_COUNT  \
0  0,0,96,29,0,570,505,2,57,0,24,20,0,46318,367,0...   
1                           7,103,4055,0,56,2323,0,4   
2  0,0,149,494,0,0,1857,0,33,269,25,1337,0,0,0,0,0,0   
3       0,15,0,98,521,0,7,0,31,100,3904,0,0,0,0,17,0   
4  0,9602,169,104,0,1588,0,0,3,0,0,114,42,0,23,17...   

                                    CODE_LINES_COUNT  \
0  0,0,299,367,0,2898,2154,238,336,16514,138,48,0...   
1                    65,538,20410,2031,102,4865,0,10   
2  0,0,2320,4056,0,14,9595,224,147,958,83,4739,0,...   
3  0,56,0,440,1417,0,16,506,116,333,12869,0,0,0,0...   
4  0,39888,912,613,0,5342,0,0,11,0,846,380,104,0,...   

                                 COMMENT_LINES_COUNT  \
0  0,0,10,79,0,726,3,0,140,0,9,0,0,69211,319,0,0,...   
1                            1,32,13156,0,0,2108,0,8   
2       0,0,60,46,0,0,2246,0,5,0,29,2444,0,0,0,0,0,0   
3          0,9,0,15,1031,0,0,0,5,0,4268,0,0,0,0,27,0   
4  0,15353,428,193,0,1752,0,0,1,0,0,66,0,0,9,2071...   

                                             LICENSE  
0  BSD-3-Clause:0.971,BSD-3-Clause-No-Nuclear-Lic...  
1                         BSD-2-Clause-FreeBSD:0.841  
2                                                NaN  
3  deprecated_LGPL-3.0+:0.996,deprecated_LGPL-3.0...  
4                                 BSD-3-Clause:0.981

As we can see above, PGA filtered 2605 repos that has Jupyter Notebook, in line with what we saw when analyzing the index file (PGA index study notebook).

Now let's see how many siva files there are.

In theory, one repo = one siva files, but there can be more than one siva file per repo if there are completely independent branches.

To get this number, we'll examine the column B that corresponds to "SIVA_FILENAMES"

In [33]:
count_sivafiles = 0

for row in df['SIVA_FILENAMES']:
    num_sivafiles = row.split(",")
    count_sivafiles += len(num_sivafiles)

count_sivafiles

6349

Hm, this is odd. We can see that there are 2,605 repos, but 6,349 siva files. This is not normal at all.

Let's understand what's going on.

In [96]:
list_sivafiles = []

for row in df['SIVA_FILENAMES']:
    num_sivafiles = row.split(",")
    list_sivafiles.append(len(num_sivafiles))

# Checking what's the average number of siva files per repo and the standard deviation

import statistics
average = round(statistics.mean(list_sivafiles), 2)
stdev = round(statistics.stdev(list_sivafiles), 2)

print("The average number of siva files is",average, "with a standard deviation of", stdev)


The average number of siva files is 2.44 with a standard deviation of 64.53


We can see that the data is distorted by some anomaly, since a mean of 2.44 with such great standard deviation is not desireable. Let's keep digging, now visually, plotting a histogram of how the number of siva files are distributed.

In [113]:
from collections import Counter

counter_files = Counter(list_sivafiles)

print("There are:")
for c in counter_files:
    print("-", counter_files[c], "repo(s) with", c, "siva files")

There are:
- 2249 repo(s) with 1 siva files
- 301 repo(s) with 2 siva files
- 40 repo(s) with 3 siva files
- 9 repo(s) with 4 siva files
- 4 repo(s) with 5 siva files
- 1 repo(s) with 6 siva files
- 1 repo(s) with 21 siva files
- 1 repo(s) with 3295 siva files


WOWWWW found the anomaly! This **ONE** repo with 3,295 siva files on it.

Now we need to see which repo this is.

In [114]:
anomaly_sivafiles = max(list_sivafiles)

wheres_the_anomaly = [i for i, x in enumerate(list_sivafiles) if x == anomaly_sivafiles][0]

print(df.iloc[wheres_the_anomaly]['URL'])

https://github.com/google/skia-buildbot


For the sake of our current study, we'll leave this repo aside, since it's a HUGE outlier.